### PR TITLE
[active-standby] Enforce switchover based on heartbeats when mux probe keeps failing

### PR DESCRIPTION
**Pull request by zjswhhh**
### Description of PR

Summary:
Fixes # (issue)

Fixing a common issue in production. Root cause is bad firmware failing to respond to mux toggle or mux query. 

For example, in below senario, we will have two `standby` ToRs in the end, both sides will have tunnel-route installed. 
``` 
UT0: active -> toggle standby -> toggles succeeds -> healthy 
LT0: standby > receive heartbeat event active due to peer's toggle -> probe mux state -> mux state unknown -> probe mux state -> repeating ...
```
The traffic loss in this scenario is evitable, as heartbeat behavior is very consistent and from that linkmgrd should be table to tell the current mux direction. 

The solution is when mux probing is the **ONLY** failing component, linkmgrd enforce a switchover based on heartbeat status. To ensure the heartbeat behaviour is consistent, we count how many times we enter the composite state **continuously**. 

sign-off: Jing Zhang zhangjing@microsoft.com 
### Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Doc/Design
- [ ] Unit test

#### What is the motivation for this PR?
Minimize traffic loss. 

#### How did you verify/test it?
Tested on dualtor DUT, with mux-simulator off. 
1. Toggle `active`
2. Toggle `standby`

Tunnel route was fixed into "one side active, one side standby" state.

### CODE REVIEW

File: src/link_manager/LinkManagerStateMachineActiveStandby.cpp [hunk 1]
```diff
@@ -1158,10 +1174,14 @@ void ActiveStandbyStateMachine::LinkProberStandbyMuxUnknownLinkUpTransitionFunct
 {
     MUXLOGINFO(mMuxPortConfig.getPortName());
 
+    mStandbyUnknownUpCount++;
+
     if ((ps(mCompositeState) != ps(nextState)) &&
         (ps(nextState) == link_prober::LinkProberState::Label::Active ||
          ps(nextState) == link_prober::LinkProberState::Label::Standby)) {
         enterMuxWaitState(nextState);
+    } else if (mStandbyUnknownUpCount == mMuxPortConfig.getNegativeStateChangeRetryCount()) {
```
Comment (yxieca) [inline src/link_manager/LinkManagerStateMachineActiveStandby.cpp:1183]: Could this branch be missed if the code took the first if branch when condition is met?

Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachineActiveStandby.cpp:1183]: When it could, it should. Because there is a heartbeat state change here.

**Review by lolyu — APPROVED** [review]

File: src/DbInterface.cpp [hunk 2]
```diff
@@ -234,7 +234,7 @@ void DbInterface::handlePostSwitchCause(
         boost::posix_time::ptime time
 )
 {
-    MUXLOGDEBUG(boost::format("%s: post switch cause %s") %
+    MUXLOGWARNING(boost::format("%s: post last switch cause %s") %
```
Comment (theasianpianist) [inline src/DbInterface.cpp:237]: Is this too verbose? What's the reason for the change?

File: src/link_manager/LinkManagerStateMachineActiveStandby.cpp (hunk 1 above)
Comment (yxieca) [reply src/link_manager/LinkManagerStateMachineActiveStandby.cpp:1183]: Are you saying that the state change will reset the counter?

File: src/DbInterface.cpp (hunk 2 above)
Comment (zjswhhh) [reply src/DbInterface.cpp:237]: > Is this too verbose? What's the reason for the change?

It will add one extra log for each switchover. 

Telemetry won't onboard switch cause data to Kusto anytime soon, so I wanted to have it in syslog for now.

File: src/link_manager/LinkManagerStateMachineActiveStandby.cpp (hunk 1 above)
Comment (zjswhhh) [reply src/link_manager/LinkManagerStateMachineActiveStandby.cpp:1183]: It should have been reset. 

If it didn't get reset, the only case I can think of is user trigger a manual toggle in between, in that case, still we should skip the new branch.

**Comment by xumia** [comment]
It did not list all the checks in job level, caused by Azp connection settings.
image

**Comment by zjswhhh** [comment]
Hi @xumia - Thanks for checking! But now it seems to be hanging for overall status?

**Comment by qiluo-msft** [comment]
Please add ADO number in PR description.

**Comment by zjswhhh** [comment]
> Please add ADO number in PR description.

Done
